### ISSUE TITLE
WasmFS: `ioctl`

### ISSUE THREAD

**Issue by tlively** [emscripten-core/emscripten#15968]
[man page](https://man7.org/linux/man-pages/man2/ioctl.2.html) although there isn't really standard behavior, so we should just try to match the current system's ioctl.

**Comment by RReverser** [comment]
One update here is that since #16677 landed there were some TTY hooks added in #19499 to support [xterm-pty](https://xterm-pty.netlify.app/).

Those hooks cover couple of ioctl functions important for the terminal, and without some analogues it won't be possible to build libraries like xterm-pty on top of WasmFS instead of the old FS.

That said, instead of covering them one by one, could WasmFS allow to intercept arbitrary ioctl syscalls in custom device implementation? I can think of some usecases beyond the terminal, e.g. for porting [umockdev](https://github.com/martinpitt/umockdev).

**Comment by tlively** [comment]
Letting backends handle ioctls if they want to sgtm.

**Comment by kripken** [comment]
@RReverser atm no, but the general idea we've had for this is to add a bridge from WasmFS to old FS backends, https://github.com/emscripten-core/emscripten/issues/18935#issuecomment-1464197239

**Comment by RReverser** [comment]
I don't remember without looking up, does TTY have a separate backend in legacy scheme? If so, that would help a lot, yeah.

**Comment by kripken** [comment]
I believe that is `src/library_tty.js`

**Comment by RReverser** [comment]
Yeah I know, just not sure if it counts as an actual backend. I thought it's just a JS library but not really a FS backend as e.g. MEMFS or WORKERFS or others are, so it's not clear to me if "add a bridge from WasmFS to old FS backends" would solve the issue or not.

**Comment by kripken** [comment]
I believe it is a backend - it has `stream_ops` etc. But there might be special hooks in the old FS for it, which might be an issue. I guess there have to be, since it isn't manually mounted like other backends...

**Comment by RReverser** [comment]
> But there might be special hooks in the old FS for it, which might be an issue.

Yeah I suspect so too, it would need to somehow proxy at least the hooks in above mentioned https://github.com/emscripten-core/emscripten/pull/19499.
